src/wmi.c

```diff
@@ -0,0 +1,548 @@
+    pi->num_parts = num_parts;
+    return (pi);
+}
+
+void metadata_str_free (metadata_str_t *ms)
+{
+    if (!ms) return;
+
+    int i;
+    for (i = 0; i < ms->num_parts; i++)
+        free (ms->parts[i]);
+    free (ms);
+}
+
+/* WMI specific */
+typedef struct wmi_connection_s
+{
+    IDispatch *dispatcher;
+} wmi_connection_t;
+
+typedef struct wmi_result_list_s
+{
+    IDispatch* results;
+
+    int count;
+    int last_result;
+} wmi_result_list_t;
+
+typedef struct wmi_result_s
+{
+    IDispatch* result;
+} wmi_result_t;
+
+HRESULT wmi_invoke_method (IDispatch *dispatcher, const wchar_t *method_name, DISPPARAMS *params, VARIANT *result)
+{
+    HRESULT hr;
+    DISPID dispid[1];
+    BSTR name;
+
+    name = SysAllocString (method_name);
+    hr = dispatcher->lpVtbl->GetIDsOfNames (dispatcher, &IID_NULL, &name, 1,
+            LOCALE_SYSTEM_DEFAULT, dispid);
+    SysFreeString (name);
+
+    if (FAILED (hr))
+        return (hr);
+
+    hr = dispatcher->lpVtbl->Invoke (dispatcher, dispid[0], &IID_NULL,
+            LOCALE_SYSTEM_DEFAULT, DISPATCH_METHOD, params, result, NULL, NULL);
+    if (FAILED (hr))
+        return (hr);
+
+    return (0);
+}
+
+int wmi_get_property (IDispatch *dispatcher, const wchar_t *property_name,
+        VARIANT *result)
+{
+    DISPPARAMS params;
+    HRESULT hr;
+    DISPID dispid[1];
+    BSTR name;
+
+    name = SysAllocString (property_name);
+    hr = dispatcher->lpVtbl->GetIDsOfNames (dispatcher, &IID_NULL,
+            &name, 1, LOCALE_SYSTEM_DEFAULT, dispid);
+    SysFreeString(name);
+
+    if (FAILED (hr))
+      return (hr);
+
+    params.cArgs = 0;
+    params.cNamedArgs = 0;
+
+    hr = dispatcher->lpVtbl->Invoke (dispatcher, dispid[0], &IID_NULL,
+            LOCALE_SYSTEM_DEFAULT, DISPATCH_PROPERTYGET, &params, result, NULL, NULL);
+    if (FAILED (hr))
+      return (hr);
+
+    return (0);
+}
+
+wmi_result_list_t* wmi_query (wmi_connection_t *connection, const wchar_t *query)
+{
+    HRESULT status = 0;
+    wmi_result_list_t *res = NULL;
+    const char *error_details = NULL;
+
+    VARIANT result;
+    VARIANTARG args[1];
+    DISPPARAMS params;
+
+    params.cNamedArgs = 0;
+    params.rgvarg = args;
+
+    params.cArgs = 1;
+    params.rgvarg[0].vt = VT_BSTR;
+    params.rgvarg[0].bstrVal = SysAllocString(query);
+
+    status = wmi_invoke_method (connection->dispatcher, L"ExecQuery", &params, &result);
+    SysFreeString (params.rgvarg[0].bstrVal);
+    if (FAILED (status))
+    {
+        error_details = "ExecQuery() failed.";
+        goto err;
+    }
+
+    res = malloc (sizeof (wmi_result_list_t));
+    res->results = result.pdispVal;
+
+    status = wmi_get_property (res->results, L"Count", &result);
+    if (FAILED(status))
+    {
+        // It is very likely that the set returned by ExecQuery is empty,
+        // which seems to be signaled by a missing 'Count' property.
+        // Because of that, we're not going to treat it as an error.
+        result.intVal = 0;
```

Comment:
I would instead assign res->count = 0 here and add an "else" to assign res->count = result.intVal...